core/src/main/java/org/apache/iceberg/BaseFile.java

```diff
@@ -185,13 +190,25 @@ public PartitionData copy() {
     this.partitionType = toCopy.partitionType;
     this.recordCount = toCopy.recordCount;
     this.fileSizeInBytes = toCopy.fileSizeInBytes;
-    if (fullCopy) {
-      this.columnSizes = SerializableMap.copyOf(toCopy.columnSizes);
-      this.valueCounts = SerializableMap.copyOf(toCopy.valueCounts);
-      this.nullValueCounts = SerializableMap.copyOf(toCopy.nullValueCounts);
-      this.nanValueCounts = SerializableMap.copyOf(toCopy.nanValueCounts);
-      this.lowerBounds = SerializableByteBufferMap.wrap(SerializableMap.copyOf(toCopy.lowerBounds));
-      this.upperBounds = SerializableByteBufferMap.wrap(SerializableMap.copyOf(toCopy.upperBounds));
+    if (copyStats) {
```

Comment:
In my view, this piece should not use streams, closures and must do as few extra copies as possible. It is called in a tight loop and our previous profiling showed that all this cost adds up at scale.

I would suggest extending `SerializableMap` with `filteredCopyOf`.

```
public static <K, V> SerializableMap<K, V> filteredCopyOf(Map<K, V> map, Set<K> keys) {
  return map == null ? null : new SerializableMap<>(map, keys);
}

private SerializableMap(Map<K, V> map, Set<K> keys) {
  Map<K, V> filteredMap = Maps.newHashMapWithExpectedSize(keys.size());

  for (K key : keys) {
    V value = map.get(key);
    if (value != null) {
      filteredMap.put(key, value);
    }
  }

  this.copiedMap = filteredMap;
}
```

Then we can define some extra helper methods in `BaseFile` to have reasonable formatting.

```
private static <K, V> Map<K, V> copyMap(Map<K, V> map, Set<K> keys) {
  if (keys == null) {
    return SerializableMap.copyOf(map);
  } else {
    return SerializableMap.filteredCopyOf(map, keys);
  }
}

private static Map<Integer, ByteBuffer> copyByteBufferMap(
    Map<Integer, ByteBuffer> map, Set<Integer> keys) {
  return SerializableByteBufferMap.wrap(copyMap(map, keys));
}
```

I understand we are trying to be generic here but I think the performance is critical.

Reply:
I went back to my original solution (the most efficient one), and used the `SerializableMap` changes you have suggested.

Also changed a bit of the `SerializableMap` constructor to address @stevenzwu previous comments.
In case the source map contains `null` values, then we do not copy them to the target map. We should use the following code to correctly handle `null` values:
```
    for (K key : keys) {
      if (map.containsKey(key)) {
        filteredMap.put(key, map.get(key));
      }
    }
```

Reply:
Had to move the proposed `SerializableMap` changes back to the `BaseFile` to avoid changing the default serialization of the `SerializableMap`:
```
    * Just this break:
        ./gradlew :iceberg-core:revapiAcceptBreak --justification "{why this break is ok}" \
          --code "java.class.defaultSerializationChanged" \
          --old "class org.apache.iceberg.util.SerializableMap<K extends java.lang.Object, V extends java.lang.Object>" \
          --new "class org.apache.iceberg.util.SerializableMap<K extends java.lang.Object, V extends java.lang.Object>"
```